python/trainers/buffer.py

```diff
@@ -0,0 +1,233 @@
+				"""
+				#TODO: Decide what to do if there enough points to retrieve 
+				if training_length is None:
+					if batch_size is None:
+						#return all of them
+						return np.array(self)
+					else:
+						# return the batch_size last elements
+						if batch_size > len(self):
+							raise BufferException("Batch size requested is too large")
+						return np.array(self[-batch_size:])
+				else:
+					if not sequential:
+						if batch_size is None:
+							# retrieve the maximum number of elements
+							batch_size = len(self) - training_length + 1
+						if (len(self) - training_length + 1) < batch_size :
+							raise BufferException("The batch size and training length requested for get_batch where" 
+								" too large given the current number of data points.")
+							return 
+						tmp_list = []
+						for end in range(len(self)-batch_size+1, len(self)+1):
+							tmp_list += [np.array(self[end-training_length:end])]
+						return np.array(tmp_list)
+					if sequential:
+						# No padding at all
+						leftover = len(self) % training_length
+						if batch_size is None:
+							# retrieve the maximum number of elements
+							batch_size = len(self) // training_length +1 *(leftover != 0)
+						if batch_size > (len(self) // training_length +1 *(leftover != 0)):
+							raise BufferException("The batch size and training length requested for get_batch where" 
+								" too large given the current number of data points.")
+							return 
+						tmp_list = []
+						padding = np.array(self[-1]) * 0 
+						for end in range(len(self), len(self) % training_length , -training_length)[:batch_size]:
+							tmp_list += [np.array(self[end-training_length:end])]
+						if (leftover != 0) and (len(tmp_list) < batch_size):
+							tmp_list +=[np.array([padding]*(training_length - leftover)+self[:leftover])]
+						return np.array(tmp_list)
+					# TODO: decide if we need to reset the local buffer now ?	
+			def reset_field(self):
+				"""
+				Resets the AgentBufferField
+				"""
+				self[:] = []
+		def __str__(self):
+			return ", ".join(["'{0}' : {1}".format(k, str(self[k])) for k in self.keys()])
+		def reset_agent(self):
+			"""
+			Resets the AgentBuffer
+			"""
+			#TODO: There might be some garbage collection issues ?
+			for k in self.keys():
+				try:
+					self[k].reset_field()
+				except:
+					print(k)
+		def __getitem__(self, key):
+			if key not in self.keys():
+				self[key] = self.AgentBufferField()
+			return super(Buffer.AgentBuffer, self).__getitem__(key)
+		def check_length(self, key_list):
+			"""
+			Some methods will require that some fields have the same length.
+			check_length will return true if the fields in key_list 
+			have the same length.
+			:param key_list: The fields which length will be compared
+			"""
+			if len(key_list) < 2:
+				return True
+			l = None
+			for key in key_list:
+				if key not in self.keys():
+					return False
+				if ((l != None) and (l!=len(self[key]))):
```

Comment:
you can omit the left side of the and - it's always safe to compare `l` to the output of `len(self[key])` since that call can never return `None`